File: occasum-core/src/main/java/pw/stamina/occasum/node/StandardPropertyNodeFactory.java
package pw.stamina.occasum.node;

import pw.stamina.occasum.PropertyHandle;
import pw.stamina.occasum.node.factory.AbstractPropertyNodeFactory;
import pw.stamina.occasum.properties.Property;

import javax.inject.Inject;

public final class StandardPropertyNodeFactory extends AbstractPropertyNodeFactory {

    @Inject
    StandardPropertyNodeFactory() {}

    @Override
    protected PropertyNode createRoot(PropertyHandle handle) {
        return new RootPropertyNode(this, handle);
    }

    @Override
    protected PropertyNode createProperty(PropertyHandle handle, PropertyNode parent, Property property) {
        return new SimplePropertyNode(this, handle, parent, property);
    }

    @Override
    protected PropertyNode createFolder(PropertyHandle handle, PropertyNode parent, String name) {
        return new FolderPropertyNode(this, handle, parent, name);
    }

    public static StandardPropertyNodeFactory newInstance() {
        return new StandardPropertyNodeFactory();
    }
}

File: occasum-core/src/main/kotlin/pw/stamina/occasum/node/StandardPropertyNodeFactory.kt
package pw.stamina.occasum.node

import pw.stamina.occasum.PropertyHandle
import pw.stamina.occasum.node.factory.AbstractPropertyNodeFactory
import pw.stamina.occasum.properties.Property

import javax.inject.Inject

class StandardPropertyNodeFactory @Inject internal constructor() : AbstractPropertyNodeFactory() {

    override fun createRoot(handle: PropertyHandle): PropertyNode {
        return RootPropertyNode(this, handle)
    }

    override fun createProperty(handle: PropertyHandle, parent: PropertyNode, property: Property): PropertyNode {
        return SimplePropertyNode(this, handle, parent, property)
    }

    override fun createFolder(handle: PropertyHandle, parent: PropertyNode, name: String): PropertyNode {
        return FolderPropertyNode(this, handle, parent, name)
    }

    companion object {

        fun newInstance(): StandardPropertyNodeFactory {
            return StandardPropertyNodeFactory()
        }
    }
}
